File: roms/xevious.py
#
#	Xevious
#

from base64 import b64encode
from io import BytesIO
from math import ceil
from sys import argv
from zipfile import ZipFile
from PIL import Image

with ZipFile(argv[1]) as z:
    prg1 = z.read('xvi_1.3p') + z.read('xvi_2.3m') + z.read('xvi_3.2m') + z.read('xvi_4.2l')
    prg2 = z.read('xvi_5.3f') + z.read('xvi_6.3j')
    prg3 = z.read('xvi_7.2c')
    bg2 = z.read('xvi_12.3b')
    bg4 = z.read('xvi_13.3c') + z.read('xvi_14.3d')
    obj = z.read('xvi_15.4m') + z.read('xvi_17.4p') + z.read('xvi_18.4r') + z.read('xvi_16.4n')
    maptbl = z.read('xvi_9.2a') + z.read('xvi_10.2b')
    mapdata = z.read('xvi_11.2c')
    red = z.read('xvi-8.6a')
    green = z.read('xvi-9.6d')
    blue = z.read('xvi-10.6e')
    bgcolor_l = z.read('xvi-7.4h')
    bgcolor_h = z.read('xvi-6.4f')
    objcolor_l = z.read('xvi-4.3l')
    objcolor_h = z.read('xvi-5.3m')
    snd = z.read('xvi-2.7n')
with ZipFile(argv[2]) as z:
    key = z.read('50xx.bin')
with ZipFile(argv[3]) as z:
    io = z.read('51xx.bin')
with ZipFile(argv[4]) as z:
    prg = z.read('54xx.bin')

rom = prg1 + prg2 + prg3 + bg2 + bg4 + obj + maptbl + mapdata + red + green + blue + bgcolor_l + bgcolor_h + objcolor_l + objcolor_h + snd + key + io + prg

def pngstring(a):
    w = 1024
    img = Image.new('P', (w, ceil(len(a) / w)))
    img.putpalette(sum([[i, 0, 0] for i in range(256)], []))
    img.putdata(a)
    buf = BytesIO()
    img.save(buf, 'PNG')
    return b64encode(buf.getvalue())

with open(argv[5], 'wb') as f:
    f.write(b'export const ROM = \'data:image/png;base64,' + pngstring(rom) + b'\';\n')

File: roms/xevious.js
/*
 *	Xevious
 */

const fs = require('fs');
const {createCanvas, createImageData} = require('canvas');
const {Zlib} = require('zlibjs/bin/unzip.min');
Zlib.Unzip.prototype.read = Zlib.Unzip.prototype.decompress;

const cat = (...args) => {
	const array = new Uint8Array(args.reduce((a, b) => a + b.length, 0));
	return args.reduce((a, b) => (array.set(b, a), a + b.length), 0), array;
};

let z = new Zlib.Unzip(fs.readFileSync(process.argv[2]));
const prg1 = cat(z.read('xvi_1.3p'), z.read('xvi_2.3m'), z.read('xvi_3.2m'), z.read('xvi_4.2l'));
const prg2 = cat(z.read('xvi_5.3f'), z.read('xvi_6.3j'));
const prg3 = z.read('xvi_7.2c');
const bg2 = z.read('xvi_12.3b');
const bg4 = cat(z.read('xvi_13.3c'), z.read('xvi_14.3d'));
const obj = cat(z.read('xvi_15.4m'), z.read('xvi_17.4p'), z.read('xvi_18.4r'), z.read('xvi_16.4n'));
const maptbl = cat(z.read('xvi_9.2a'), z.read('xvi_10.2b'));
const mapdata = z.read('xvi_11.2c');
const red = z.read('xvi-8.6a');
const green = z.read('xvi-9.6d');
const blue = z.read('xvi-10.6e');
const bgcolor_l = z.read('xvi-7.4h');
const bgcolor_h = z.read('xvi-6.4f');
const objcolor_l = z.read('xvi-4.3l');
const objcolor_h = z.read('xvi-5.3m');
const snd = z.read('xvi-2.7n');
z = new Zlib.Unzip(fs.readFileSync(process.argv[3]));
const key = z.read('50xx.bin');
z = new Zlib.Unzip(fs.readFileSync(process.argv[4]));
const io = z.read('51xx.bin');
z = new Zlib.Unzip(fs.readFileSync(process.argv[5]));
const prg = z.read('54xx.bin');

const rom = cat(prg1, prg2, prg3, bg2, bg4, obj, maptbl, mapdata, red, green, blue, bgcolor_l, bgcolor_h, objcolor_l, objcolor_h, snd, key, io, prg);

const pngString = array => {
	const w = 1024, h = Math.ceil(array.byteLength / w), canvas = createCanvas(w, h);
	const palette = new Uint8ClampedArray(new Uint32Array(256).map((e, i) => i | 0xff000000).buffer);
	canvas.getContext('2d', {pixelFormat: 'A8'}).putImageData(createImageData(new Uint8ClampedArray(array.buffer), w, h), 0, 0);
	return Buffer.from(canvas.toBuffer('image/png', {compressionLevel: 9, filters: canvas.PNG_FILTER_NONE, palette})).toString('base64');
};

fs.writeFileSync(process.argv[6], `export const ROM = 'data:image/png;base64,${pngString(rom)}';\n`);
